src/Traversals/depthfirst.jl

```diff
@@ -19,44 +19,46 @@ function traverse_graph!(
 
     n = nv(g)
     visited = falses(n)
-    S = Vector{Tuple{U,U}}()
+    S = Vector{Tuple{U, U}}()
     sizehint!(S, length(ss))
     preinitfn!(state, visited) || return false
     @inbounds for s in ss
         us = U(s)
+        visited[us] && continue
+
         visited[us] = true
         push!(S, (us, 1))
         initfn!(state, us) || return false
-    end
 
-    ptr = one(U)
-
-    while !isempty(S)
-        v, _ = S[end]
-        previsitfn!(state, v) || return false
-
-        neighs=alg.neighborfn(g, v)
-        @inbounds while ptr <= length(neighs)
-            i = neighs[ptr]
-            visitfn!(state, v, i) || return false
-            if !visited[i]  # find the first unvisited neighbor
-                newvisitfn!(state, v, i) || return false
-                visited[i] = true
-                push!(S, (i, ptr+1))
-                break
+        ptr = one(U)
+
+        while !isempty(S)
+            v, _ = S[end]
+            previsitfn!(state, v) || return false
+
+            neighs=alg.neighborfn(g, v)
+            @inbounds while ptr <= length(neighs)
+                i = neighs[ptr]
+                visitfn!(state, v, i) || return false
+                if !visited[i]  # find the first unvisited neighbor
+                    newvisitfn!(state, v, i) || return false
+                    visited[i] = true
+                    push!(S, (i, ptr+1))
+                    break
+                end
+                ptr += 1
+            end
+            postvisitfn!(state, v) || return false
+            # if ptr > length(neighs) then we have finished all children of the node,
+            # and the next time we pop from the stack we will be in the parent of the current
+            # node. We would like to continue from where we stoped, otherwise we have found
+            # a new unvisited child, so we will make ptr = 1
+            if ptr > length(neighs)
+                postlevelfn!(state) || return false
+                _, ptr =pop!(S)
+            else
+                ptr = 1 # we will enter new node
             end
-            ptr += 1
-        end
-        postvisitfn!(state, v) || return false
-        # if ptr > length(neighs) then we have finished all children of the node,
-        # and the next time we pop from the stack we will be in the parent of the current
-        # node. We would like to continue from where we stoped, otherwise we have found
-        # a new unvisited child, so we will make ptr = 1
-        if ptr > length(neighs)
-            postlevelfn!(state) || return false
-            _, ptr =pop!(S)
-        else
-            ptr = 1 # we will enter new node
```

Comment:
what is changing in this block?

Reply:
in the old dfs, we were always push the the vertices in ss to the stack, despite that some of the theme by the time it would, be poped, it would have been visited.

Reply:
Let me be specific: what is changing between lines 32 and 59?

Reply:
nothing have changed, expect that is those lines are now in nested block.

Reply:
wouldn't it be easier to keep this as it is and have a

```
if visited[v]
    pop!(S)
    continue
end
``` 

after the inside the `while !isempty(S)` loop?

Reply:
in many cases we will have to add all the vertices to `ss` pushing all the vertices in the stack and then poping them, while result in unnecessary memory allocations.

Reply:
Yes, but it's more consistent with `initfn!` in BreadthFirst, which executes against all sources before `visitfn!`. Can you provide some benchmarks showing the inefficiency?

Reply:
there is something i Remember, initfn must be applied only after a new connected compnent  have been discovered, so we can’t do it that way.

Reply:
Why? Can't we use `previsitfn!` for that?

Reply:
for example in `strongly_connected_component ` , `previsitfn!` can’t by its own know if the node we are in, is visited before or not, so i made the variable `up` in the state, which will be true if the the node which will be visited in the next `previsitfn` have been visited before, so `s.up` is set to `false` in two cases, if we descover a new edge, or we discovered a new connected component, so we need `initfn` to be executed only before we dicover new connected component

Reply:
plus i see no harm to  make the time we visit `initfn!` is different from `BreadthFirst`, in `BFS` we visit the nodes level by level, and the sources is known in advance, so it is logical  to start by visiting all the sources, which is n’t the case in `DFS`.

Reply:
Yeah. I'm coming around to your line of thinking. It might even make sense not to allow multiple sources in DFS.

Reply:
in `dfs` there might be already multiple sources, if the graph has more than one connected component, me can’t figure that until we do the algorithm, so we might need to include all the vertices in the graph in `ss` argument
.